cli/src/commands/extensions/state.rs

```diff
@@ -0,0 +1,94 @@
+//! Shared extension state.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use anyhow::{anyhow, Result};
+use deno_runtime::deno_core::OpState;
+use futures::future::BoxFuture;
+use tokio::sync::Mutex;
+
+use crate::commands::extensions::permissions::Permissions;
+use crate::commands::extensions::PhylumApi;
+
+/// Holds either an unawaited, boxed `Future`, or the result of awaiting the
+/// future.
+enum OnceFuture<T: Unpin> {
+    Future(BoxFuture<'static, T>),
+    Awaited(T),
+}
+
+impl<T: Unpin> OnceFuture<T> {
+    fn new(inner: BoxFuture<'static, T>) -> Self {
+        OnceFuture::Future(inner)
+    }
+
+    async fn get(&mut self) -> &T {
+        match *self {
+            OnceFuture::Future(ref mut inner) => {
+                *self = OnceFuture::Awaited(inner.await);
+                match *self {
+                    OnceFuture::Future(..) => unreachable!(),
+                    OnceFuture::Awaited(ref mut inner) => inner,
+                }
+            },
+            OnceFuture::Awaited(ref mut inner) => inner,
+        }
+    }
+}
+
+// XXX: Holding a `Ref`, `RefMut`, or guard across await points, can cause
+// issues when that field is accessed from another extension API
+// method.
+//
+// If a field in [`ExtensionStateInner`] requires mutable access, it should be
+// synchronized with an async-safe `Mutex` or `RwLock`, so an attempted access
+// to the inner value will cause blocking until the initial guard is resolved.
+//
+// This also means that if this guard is held across an await point, all other
+// access will stall until that future is completed. Avoid holding guards across
+// await points, especially when depending on other methods accessing the same
+// lock.
+//
+// Using `RefCell` to avoid locking should only be done when the value is
+// private and never held across await points internally. Prefer [`Cell`] if the
+// inner value only needs to be replaced.
+//
+/// Extension state the APIs have access to.
+struct ExtensionStateInner {
+    api: Mutex<OnceFuture<Result<Rc<PhylumApi>>>>,
+    permissions: Permissions,
+}
```

Comment:
This is intentionally private without `Deref`. Unfortunately this means that access to things like `Permissions` which are not public will require a plumbing method in `ExtensionState`, but this way it's impossible to directly extract the `ExtensionStateInner` from the `OpState` without going through `ExtensionState`.

TODO(chris): Is this actually necessary? If `ExtensionState` is stored on `OpState`, typesystem shouldn't allow extracting `Rc<ExtensionStateInner>` directly.